captum/attr/_core/occlusion.py

```diff
@@ -0,0 +1,336 @@
+        being 1. This mask contains 1s in locations which have been ablated (and
+        thus counted towards ablations for that feature) and 0s otherwise.
+        """
+        input_mask = torch.stack(
+            [
+                self._occlusion_mask(
+                    expanded_input,
+                    j,
+                    kwargs["sliding_window_tensors"],
+                    kwargs["strides"],
+                    kwargs["shift_counts"],
+                )
+                for j in range(start_feature, end_feature)
+            ],
+            dim=0,
+        ).long()
+        ablated_tensor = (expanded_input * (1 - input_mask).float()) + (
+            baseline * input_mask.float()
+        )
+        return ablated_tensor, input_mask
+
+    def _occlusion_mask(
+        self,
+        expanded_input,
+        ablated_feature_num,
+        sliding_window_tsr,
+        strides,
+        shift_counts,
+    ):
+        """
+        This constructs the current occlusion mask, which is the appropriate
+        shift of the sliding window tensor based on the ablated feature number.
+        The feature number ranges between 0 and the product of the shift counts
+        (# of times the sliding window should be shifted in each dimension).
+
+        First, the ablated feature number is converted to the number of steps in
+        each dimension from the origin, based on shift counts. This procedure
+        is similar to a base conversion, with the position values equal to shift
+        counts. The feature number is first taken modulo shift_counts[0] to
+        get the number of shifts in the first dimension (each shift
+        by shift_count[0]), and then divided by shift_count[0].
+        The procedure is then continued for each element of shift_count. This
+        computes the total shift in each direction for the sliding window.
+
+        We then need to compute the padding required after the window in each
+        dimension, which is equal to the total input dimension minus the sliding
+        window dimension minus the (left) shift amount. We construct the
+        array pad_values which contains the left and right pad values for each
+        dimension, in reverse order of dimensions, starting from the last one.
+
+        Once these padding values are computed, we pad the sliding window tensor
+        of 1s with 0s appropriately, which is the corresponding mask,
+        and the result will match the input shape.
+        """
+        remaining_total = ablated_feature_num
+        current_index = []
+        for i, shift_count in enumerate(shift_counts):
+            stride = strides[i] if isinstance(strides, tuple) else strides
+            current_index.append((remaining_total % shift_count) * stride)
+            remaining_total = remaining_total // shift_count
+
+        remaining_padding = np.subtract(
+            expanded_input.shape[2:], np.add(current_index, sliding_window_tsr.shape)
```

Comment:
Can this lead to negative paddings ?

Reply:
Yes, this is intentional, since passing negative padding splices the corresponding dimension, which is used when the sliding window tensor needs to be cropped at edges.